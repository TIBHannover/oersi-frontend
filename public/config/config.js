window['runTimeConfig'] = {
  ELASTIC_SEARCH: {
    URL: "http://localhost:9200/",
    CREDENTIALS: "",
    APP_NAME: "oer_data"
  },
  GENERAL_CONFIGURATION:{
    AVAILABLE_LANGUAGES: ["de", "en"],
    PUBLIC_URL: "http://localhost/resources",
    RESULT_PAGE_SIZE_OPTIONS:["12", "24", "48", "96"],  // page size options configuration
    NR_OF_RESULT_PER_PAGE:12,  //  number of results to show per view. Defaults to 12.
    HEADER_LOGO_URL: "",  // if set, use this URL for the logo in the header and include your custom svg or similar; default is "logo-192.png". You may use placeholders {{dark}} and {{small}} to use different logo-versions for dark/mobile mode (they will be replaced by "_dark" and "_small").
    THEME_COLORS: null,  // customize colors of the theme; format: {primary: {main: "#000"}, secondary: {main: "#000"}}
    THEME_COLORS_DARK: null,  // customize colors of the dark theme; format: {primary: {main: "#000"}, secondary: {main: "#000"}}
    /**
     * Accept a list of objects 
     * example:
     * {'path': 'public/{folderName}/{languageCode}/{fileName}.html', 'language': '{languageCode}'}
     * 
     */
    PRIVACY_POLICY_LINK: [],
    I18N_CACHE_EXPIRATION: 600000, // expiration time of the i18n translation cache storage
    I18N_DEBUG: false,
    TRACK_TOTAL_HITS: true, // track number of total hits from elasticsearch - see https://www.elastic.co/guide/en/elasticsearch/reference/7.10/search-your-data.html#track-total-hits
    FEATURES: {
      DARK_MODE: true,
      CHANGE_FONTSIZE: false,  // experimental/beta - just to show weaknesses in styling
      EMBED_OER: true, // feature toggle: use "embed-oer" button
      SCROLL_TOP_BUTTON: true // feature toggle: use "scroll-to-top" button
    }
  }
}