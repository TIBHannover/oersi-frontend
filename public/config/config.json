{
  "ELASTIC_SEARCH": {
    "URL": "http://localhost:9200/",
    "CREDENCIAL": "",
    "APP_NAME": "oer_data"
  }
}
