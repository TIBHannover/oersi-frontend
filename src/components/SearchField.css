svg.search-icon {
  fill: #e86161 !important;
  height: 1.2em;
}

/* workaround: suggestion-list for "no suggestion" remains with white background in reactivesearch-dark-mode */
.search-component-dark li {
  background-color: rgb(66,66,66);
}

/* workaround: suggestion-list uses light-mode scrollbars in reactivesearch-dark-mode */
.search-component-dark ul {
  scrollbar-color: grey rgb(66,66,66);
}

.search-component-input {
  border-radius: 1em;
}

.search-component-input:focus {
  box-shadow: 0 0 0 0.2rem rgba(232, 97, 97, 0.25);
}
