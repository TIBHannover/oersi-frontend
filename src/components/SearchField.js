import React, {useState, useEffect} from "react"
import {SearchBox, useSearchBox} from "react-instantsearch"
import SearchIndexFrontendConfigContext from "../helpers/SearchIndexFrontendConfigContext"
import {useTranslation} from "next-i18next"
import {
  Autocomplete,
  Box,
  CircularProgress,
  TextField,
  useTheme,
} from "@mui/material"

const SearchField = (props) => {
  const theme = useTheme()
  const {t} = useTranslation()
  const frontendConfig = React.useContext(SearchIndexFrontendConfigContext)
  const conf = frontendConfig.searchConfiguration.searchField

  return (
    <Box
      className={
        "search-component" +
        (theme.palette.mode === "dark" ? " search-component-dark" : "")
      }
      sx={{
        fontSize: theme.typography.fontSize,
        "& .search-component-input": {
          fontSize: theme.typography.fontSize,
        },
        "& li": {
          fontSize: theme.typography.fontSize,
        },
        "& .input-group": {
          boxShadow: "unset",
        },
      }}
    >
      <SearchBoxWithSuggestions />
    </Box>
  )
  function onError(error) {
    return (
      <div>
        Something went wrong!
        <br />
        Error details
        <br />
        {error}
      </div>
    )
  }

  function onNoSuggestion() {
    return <div>No suggestions found</div>
  }
}

const SearchBoxWithSuggestions = () => {
  const [open, setOpen] = useState(false)
  const [options, setOptions] = useState([])
  const [loading, setLoading] = useState(false)
  const {query, refine} = useSearchBox()
  const [inputValue, setInputValue] = useState(query)

  function setQuery(newQuery) {
    setInputValue(newQuery)
    refine(newQuery)
  }

  useEffect(() => {
    let active = true

    if (!loading) {
      return undefined
    }

    // ;(async () => {
    //   const response = await axios.get("/api/suggestions") // Replace with your API endpoint
    //   if (active) {
    //     setOptions(response.data)
    //   }
    // })()
    setOptions([
      {name: "GitHub"},
      {name: "GitLab"},
      {name: "Bitbucket"},
      {name: "GitKraken"},
      {name: "Git"},
    ])
    setLoading(false)

    return () => {
      active = false
    }
  }, [loading])

  useEffect(() => {
    if (!open) {
      setOptions([])
    }
  }, [open])

  const handleKeyDown = (event) => {
    if (event.key === "Enter") {
      event.preventDefault()
      event.defaultMuiPrevented = true
      // Handle the submission logic here
      // console.log("Submitted value:", event.target.value)
      // You can call a function to handle the submission
      setQuery(event.target.value)
    }
  }

  return (
    <Autocomplete
      id="search-box"
      freeSolo
      onKeyDown={handleKeyDown}
      open={open}
      onOpen={() => setOpen(true)}
      onClose={() => setOpen(false)}
      onInputChange={(event, value) => {
        if (value.length >= 3) {
          // Fetch suggestions when input length is 3 or more
          setLoading(true)
        } else {
          setLoading(false)
        }
      }}
      getOptionLabel={(option) => option.name} // Adjust according to your data structure
      options={options}
      loading={loading}
      renderInput={(params) => (
        <TextField
          {...params}
          label="Search"
          variant="outlined"
          // onKeyDown={handleKeyDown}
          InputProps={{
            ...params.InputProps,
            endAdornment: (
              <>
                {loading ? <CircularProgress color="inherit" size={20} /> : null}
                {params.InputProps.endAdornment}
              </>
            ),
          }}
        />
      )}
    />
  )
}

export default SearchField
