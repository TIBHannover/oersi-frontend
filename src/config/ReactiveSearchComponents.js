import React from "react"
import {
  DataSearch,
  MultiList,
  ReactiveList,
  ResultCard,
} from "@appbaseio/reactivesearch"

export default {
  datasearch: {
    className: "search",
    componentId: "search",
    dataField: ["name", "creator.name", "description", "keywords"],
    fieldWeights: [1, 3],
    queryFormat: "and",
    fuzziness: 0,
    debounce: 100,
    autosuggest: true,
    highlight: true,
    highlightField: "keywords",
    iconPosition: "right",
    URLParams: true,
    react: {
      and: ["about", "results"],
    },
    source: DataSearch,
  },
  resultcard: {
    className: "right-col",
    componentId: "results",
    dataField: "name",
    size: 12,
    render: ({data}) => (
      <ReactiveList.ResultCardsWrapper>
        {data.map((item) => (
          <ResultCard href={item._id} target="_self" key={item._id}>
            <ResultCard.Image src={item.image} />
            <ResultCard.Title>{item.name}</ResultCard.Title>
          </ResultCard>
        ))}
      </ReactiveList.ResultCardsWrapper>
    ),
    pagination: true,
    URLParams: true,
    react: {
      and: ["about", "search"],
    },
    innerClass: {
      resultStats: "result-stats",
      list: "list",
      listItem: "list-item",
      image: "image",
    },
    source: ReactiveList,
  },
  multiList: [
    {
      componentId: "about",
      key: "about",
      dataField: "about.id",
      title: "about",
      placeholder: "about",
      filterLabel: "ABOUT",
      queryFormat: "and",
      showMissing: true,
      missingLabel: "N/A",
      showFilter: true,
      showSearch: false,
      size: 1000,
      className: "about-card",
      fontAwesome: "",
      URLParams: true,
      and: ["search"],
      allowedSearchRegex: /^[\u00C0-\u017Fa-zA-Z .-]*$/, // allow only search-terms matching this regex
      source: MultiList,
    },
  ],
}
