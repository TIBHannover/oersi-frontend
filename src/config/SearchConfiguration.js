function prepareSearchConfiguration(oersiConfig) {
  const searchConfig = oersiConfig.search
  const fieldsOptions = oersiConfig.fieldConfiguration?.options
  return enrichDefaultConfig(
    {
      resultList: {
        componentId: "results",
        pagination: true,
        paginationAt: "bottom",
        ...searchConfig.resultList,
      },
      searchField: {
        componentId: "search",
        iconPosition: "right",
        ...searchConfig.searchField,
      },
      filters: searchConfig.filters,
    },
    fieldsOptions
  )
}
function enrichDefaultConfig(defaultConfig, fieldsOptions) {
  const filters = defaultConfig.filters.map((e) => ({
    componentId: e.dataField,
    ...e,
  }))
  const allComponentIds = [
    defaultConfig.resultList.componentId,
    defaultConfig.searchField.componentId,
  ].concat(filters.map((e) => e.componentId))
  const addReact = (component) => {
    return {
      ...component,
      react: {and: allComponentIds.filter((id) => id !== component.componentId)},
    }
  }
  const addCustomQuery = (component) => {
    const fieldOptions = fieldsOptions?.find(
      (e) => e.dataField === component.dataField
    )
    if (fieldOptions?.grouping) {
      return {
        ...component,
        ...getCustomAggregationQueries(
          component.dataField,
          fieldOptions?.grouping,
          fieldOptions?.collectOthersInSeparateGroup,
          component.showMissing
        ),
      }
    }
    return component
  }
  return {
    ...defaultConfig,
    resultList: addReact(defaultConfig.resultList),
    searchField: addReact(defaultConfig.searchField),
    filters: filters.map(addReact).map(addCustomQuery),
  }
}
// CustomAggregationQueries are currently only used for the license field to be able to use groups for licenses
function getCustomAggregationQueries(
  fieldName,
  groupingConfigs,
  collectOthersInSeparateGroup,
  showMissing
) {
  let aggsScript = "if (doc['" + fieldName + "'].size()==0) { return null }"
  const aggsScriptEntry = (groupingConfig) => {
    return (
      " else if (doc['" +
      fieldName +
      "'].value?.toLowerCase() =~ /" +
      groupingConfig.regex.toLowerCase().replaceAll("/", "\\/") +
      "/) { return '" +
      groupingConfig.id +
      "'}"
    )
  }
  const getGroupSearch = (v) => {
    if (showMissing && v === "N/A") {
      return {bool: {must_not: {exists: {field: fieldName}}}}
    }
    const config = groupingConfigs.find((c) => c.id === v)
    if (!config && collectOthersInSeparateGroup) {
      return {
        bool: {
          must: {exists: {field: fieldName}},
          must_not: groupingConfigs.map((c) => ({
            regexp: {[fieldName]: {value: c.regex}},
          })),
        },
      }
    }
    return {
      regexp: {
        [fieldName]: {
          value: config ? config.regex : v,
        },
      },
    }
  }
  aggsScript += groupingConfigs.reduce(
    (result, groupingConfig) => result + aggsScriptEntry(groupingConfig),
    ""
  )
  if (collectOthersInSeparateGroup) {
    aggsScript += " else { return 'OTHER' }"
  } else {
    aggsScript += " else { return doc['" + fieldName + "'] }"
  }
  const aggsQuery = {
    aggs: {
      [fieldName]: {
        terms: {
          size: 100,
          script: {
            source: aggsScript,
            lang: "painless",
          },
        },
      },
    },
  }
  if (showMissing) {
    aggsQuery["aggs"][fieldName]["terms"]["missing"] = "N/A"
  }
  return {
    defaultQuery: () => aggsQuery,
    customQuery: (value, props) => {
      return value instanceof Array
        ? {
            query: {
              bool: {
                should: [...value.map((v) => getGroupSearch(v))],
              },
            },
          }
        : {}
    },
  }
}
export default prepareSearchConfiguration
