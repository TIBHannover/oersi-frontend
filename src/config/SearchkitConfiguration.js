import {history} from "instantsearch.js/es/lib/routers"
import Client from "@searchkit/instantsearch-client"
import Searchkit from "searchkit"

export function getSearchkitRouting(
  searchIndexName,
  searchConfiguration,
  defaultPageSize
) {
  return {
    stateMapping: {
      stateToRoute(uiState) {
        const indexUiState = uiState[searchIndexName]
        let routeState = {
          search: indexUiState.query,
        }
        routeState = searchConfiguration.filters.reduce((acc, filter) => {
          if (filter.type === "switch") {
            acc[filter.componentId] = indexUiState.toggle?.[filter.componentId]
          } else {
            acc[filter.componentId] =
              indexUiState.refinementList?.[filter.componentId]
          }
          return acc
        }, routeState)
        if (
          indexUiState.hitsPerPage &&
          indexUiState.hitsPerPage !== defaultPageSize
        ) {
          routeState.size = indexUiState.hitsPerPage
        }
        if (indexUiState.page) {
          routeState.page = indexUiState.page
        }
        return routeState
      },
      routeToState(routeState) {
        const refinementList = searchConfiguration.filters
          .filter((filter) => filter.type !== "switch")
          .reduce((acc, filter) => {
            acc[filter.componentId] = routeState[filter.componentId]
            return acc
          }, {})
        const toggle = searchConfiguration.filters
          .filter((filter) => filter.type === "switch")
          .reduce((acc, filter) => {
            acc[filter.componentId] = routeState[filter.componentId]
            return acc
          }, {})
        return {
          [searchIndexName]: {
            query: routeState.search,
            refinementList: refinementList,
            toggle: toggle,
            hitsPerPage: routeState.size,
            page: routeState.page,
          },
        }
      },
    },
    router: history({
      cleanUrlOnDispose: false,
      createURL({qsModule, routeState, location}) {
        const searchParams = new URLSearchParams(location.search)
        Object.keys(routeState).forEach((key) => {
          searchParams.delete(key)
          if (routeState[key] !== undefined) {
            searchParams.set(key, encode(routeState[key]))
          }
        })
        if (!routeState.size) {
          searchParams.delete("size")
        }
        if (!routeState.page) {
          searchParams.delete("page")
        }
        const newSearch = searchParams.toString()
        return (
          location.origin +
          location.pathname +
          (newSearch ? "?" + newSearch : newSearch)
        )
      },
      parseURL({qsModule, location}) {
        const routeState = {}
        const searchParams = new URLSearchParams(location.search)
        searchParams.keys().forEach((key) => {
          const value = searchParams.get(key)
          try {
            routeState[key] = decode(value)
          } catch (e) {
            // Do not set value if JSON parsing fails.
            console.warn(`Failed to decode value for key "${key}":`, e)
          }
        })
        return routeState
      },
    }),
  }
}
function encode(value) {
  return JSON.stringify(value)
}
function decode(value) {
  return JSON.parse(value)
}

export function getSearchkitClient(backendSearchApiUrl, searchConfiguration) {
  return Client(
    new Searchkit(
      {
        connection: {
          host: backendSearchApiUrl,
        },
        search_settings: {
          search_attributes: searchConfiguration.searchField.searchAttributes,
          facet_attributes: searchConfiguration.facet_attributes,
        },
      },
      {debug: true}
    ),
    {
      getBaseFilters: () => {
        return searchConfiguration.globalDataRestrictions
      },
      getQuery: (query, search_attributes) => {
        const fields = search_attributes.map((attribute) => {
          const w = attribute.weight ? "^" + attribute.weight : ""
          return typeof attribute === "string" ? attribute : attribute.field + w
        })
        const types = ["cross_fields", "phrase", "phrase_prefix"]
        return {
          bool: {
            must: {
              bool: {
                should: types.map((type) => ({
                  multi_match: {
                    query,
                    fields: fields,
                    type: type,
                    operator: "and",
                  },
                })),
              },
            },
          },
        }
      },
      hooks: {
        beforeSearch: async (searchRequests) => {
          return searchRequests.map((sr) => {
            return {
              ...sr,
              body: {
                ...sr.body,
                track_total_hits: true,
              },
            }
          })
        },
      },
    }
  )
}
