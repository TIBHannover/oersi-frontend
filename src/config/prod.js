export default {
  name: "production",
  resultList: {
    component: "results",
    dataFiled: "name",
    stream: true,
    pagination: true,
    showResultStats: true,
    paginationAt: "bottom",
    sortBy: null,
    loader: "Fetching Data ........",
    pagesShow: 5,
    sizeShow: 5,
    URLParams: true,
    showEndPage: true,
    and: [
      "author",
      "license",
      "search",
      "provider",
      "results",
      "learningResourceType",
      "language",
    ],
    sortByDynamic: null,
  },
  searchComponent: {
    component: "search",
    dataField: ["name", "creator.name", "description"],
    fieldWeights: [1, 3],
    queryFormat: "or",
    fuzziness: 0,
    debounce: 100,
    autosuggest: true,
    highlight: true,
    highlightField: "keywords",
    iconPosition: "right",
    showFilter: true,
    URLParams: true,
    and: [
      "author",
      "license",
      "search",
      "provider",
      "results",
      "learningResourceType",
      "language",
    ],
  },
  multiList: [
    {
      component: "author",
      dataField: "creator.name.keyword",
      nestedField: "",
      title: "Autor",
      placeholder: "Autor",
      filterLabel: "Autor",
      showSearch: true,
      showFilter: true,
      className: "author-card",
      fontAwesome: "",
      URLParams: true,
      and: [
        "license",
        "search",
        "provider",
        "results",
        "learningResourceType",
        "language",
      ],
    },
    {
      component: "license",
      dataField: "license",
      title: "Lizenz",
      placeholder: "Lizenz",
      filterLabel: "Lizenz",
      showFilter: true,
      showSearch: false,
      className: "license-card",
      fontAwesome: "",
      URLParams: true,
      and: [
        "author",
        "search",
        "provider",
        "results",
        "learningResourceType",
        "language",
      ],
    },
    {
      component: "provider",
      dataField: "mainEntityOfPage.basedOn.provider",
      title: "Quelle",
      placeholder: "Quelle",
      filterLabel: "Quelle",
      showFilter: true,
      showSearch: false,
      className: "sourse-card",
      fontAwesome: "",
      URLParams: true,
      and: [
        "author",
        "license",
        "search",
        "results",
        "learningResourceType",
        "language",
      ],
    },
    {
      component: "language",
      dataField: "inLanguage",
      title: "Sprache",
      placeholder: "Sprache",
      filterLabel: "Sprache",
      showFilter: true,
      showSearch: false,
      className: "language-card",
      fontAwesome: "",
      URLParams: true,
      and: [
        "author",
        "license",
        "search",
        "provider",
        "results",
        "learningResourceType",
      ],
    },
    {
      component: "learningResourceType",
      dataField: "learningResourceType.id",
      title: "Material",
      placeholder: "Material",
      filterLabel: "Material",
      showFilter: true,
      showSearch: false,
      className: "source-type-card",
      fontAwesome: "",
      URLParams: true,
      and: ["author", "license", "search", "provider", "results", "language"],
    },
  ],
}
