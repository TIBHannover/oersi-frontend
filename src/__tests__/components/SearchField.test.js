import React from "react"
import {shallow} from "../../setupTests"
import {registerConfiguration} from "../../config/configurationData"
import SearchField from "../../components/SearchField"

describe("SearchField ==> Test UI  ", () => {
  registerConfiguration()
  const wrapperShadow = shallow(<SearchField />)
  it("SearchField : should render correctly", () => {
    expect(wrapperShadow).toMatchSnapshot()
  })
})
