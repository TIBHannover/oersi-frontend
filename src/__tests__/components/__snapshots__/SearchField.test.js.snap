// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchField ==> Test UI   SearchField : should render correctly 1`] = `ShallowWrapper {}`;
